### PR TITLE
feat: fees transactions order in mempool

### PR DESCRIPTION
**Pull request by Romsters**
# What :computer: 
- Added CLI param `--order` for transactions order in mempool.
- Added `fees` transactions order option, sorts transactions by `max_fee_per_gas`.
- FIFO is still the default option to reflect mainnet.

# Why :hand:
Feature parity with anvil.

### CODE REVIEW

File: crates/cli/src/cli.rs [hunk 1]
```diff
@@ -251,6 +251,10 @@ pub struct Cli {
     /// Disable CORS.
     #[arg(long, default_missing_value = "true", num_args(0..=1), conflicts_with = "allow_origin", help_heading = "Server options")]
     pub no_cors: Option<bool>,
+
+    /// How transactions are sorted in the mempool.
```
Comment (dutterbutter) [inline crates/cli/src/cli.rs:255]:
```suggestion
    /// Transaction ordering in the mempool.
```

File: crates/config/src/config.rs [hunk 2]
```diff
@@ -115,6 +115,8 @@ pub struct TestNodeConfig {
     pub allow_origin: String,
     /// Disable CORS if true
     pub no_cors: bool,
+    /// How transactions are sorted in the mempool
+    pub transactions_order: TransactionOrder,
```
Comment (dutterbutter) [inline crates/config/src/config.rs:119]:
```suggestion
    pub transaction_order: TransactionOrder,
```

File: crates/cli/src/main.rs [hunk 3]
```diff
@@ -276,7 +276,7 @@ async fn main() -> anyhow::Result<()> {
 
     let time = TimestampManager::default();
     let impersonation = ImpersonationManager::default();
-    let pool = TxPool::new(impersonation.clone());
+    let pool = TxPool::new(impersonation.clone(), config.transactions_order);
```
Comment (dutterbutter) [inline crates/cli/src/main.rs:279]:
```suggestion
    let pool = TxPool::new(impersonation.clone(), config.transaction_order);
```

File: crates/config/src/config.rs [hunk 4]
```diff
@@ -175,6 +177,7 @@ impl Default for TestNodeConfig {
             no_mining: false,
 
             max_transactions: 1000,
+            transactions_order: TransactionOrder::Fifo,
```
Comment (dutterbutter) [inline crates/config/src/config.rs:180]:
```suggestion
            transaction_order: TransactionOrder::Fifo,
```

File: crates/config/src/config.rs [hunk 5]
```diff
@@ -878,6 +881,13 @@ impl TestNodeConfig {
         self
     }
 
+    // Set transactions order in the mempool
```
Comment (dutterbutter) [inline crates/config/src/config.rs:884]:
```suggestion
    // Set transaction order in the mempool
```

File: crates/config/src/config.rs [hunk 6]
```diff
@@ -878,6 +881,13 @@ impl TestNodeConfig {
...
+    #[must_use]
+    pub fn with_transactions_order(mut self, transactions_order: TransactionOrder) -> Self {
+        self.transactions_order = transactions_order;
```
Comment (dutterbutter) [inline crates/config/src/config.rs:887]:
```suggestion
        self.transaction_order = transaction_order;
```

File: crates/core/src/node/in_memory.rs [hunk 7]
```diff
@@ -1263,7 +1263,10 @@ impl InMemoryNode {
         tracing::debug!(count = txs.len(), "applying transactions");
 
         // Create a temporary tx pool (i.e. state is not shared with the node mempool).
-        let pool = TxPool::new(self.impersonation.clone());
+        let pool = TxPool::new(
+            self.impersonation.clone(),
+            self.read_inner()?.config.transactions_order,
```
Comment (dutterbutter) [inline crates/core/src/node/in_memory.rs:1268]:
```suggestion
            self.read_inner()?.config.transaction_order,
```

File: crates/core/src/node/pool.rs [hunk 8]
```diff
@@ -1,59 +1,96 @@
...
 use zksync_types::l2::L2Tx;
-use zksync_types::{Address, H256};
+use zksync_types::H256;
 
 #[derive(Clone)]
 pub struct TxPool {
-    inner: Arc<RwLock<Vec<L2Tx>>>,
+    inner: Arc<RwLock<BTreeSet<PoolTransaction>>>,
+    /// How transactions are ordered in the pool
+    transactions_order: Arc<RwLock<TransactionOrder>>,
```
Comment (dutterbutter) [inline crates/core/src/node/pool.rs:14]:
```suggestion
    transaction_order: Arc<RwLock<TransactionOrder>>,
```

File: crates/core/src/node/pool.rs [hunk 9]
```diff
@@ -1,59 +1,96 @@
...
+    /// Used to preserve transactions submission order in the pool
+    submission_number: Arc<Mutex<u64>>,
     /// Listeners for new transactions' hashes
     tx_listeners: Arc<Mutex<Vec<Sender<H256>>>>,
     pub(crate) impersonation: ImpersonationManager,
 }
 
 impl TxPool {
-    pub fn new(impersonation: ImpersonationManager) -> Self {
+    pub fn new(impersonation: ImpersonationManager, transactions_order: TransactionOrder) -> Self {
```
Comment (dutterbutter) [inline crates/core/src/node/pool.rs:23]:
```suggestion
    pub fn new(impersonation: ImpersonationManager, transaction_order: TransactionOrder) -> Self {
```

File: crates/core/src/node/pool.rs [hunk 10]
```diff
@@ -1,59 +1,96 @@
...
         Self {
-            inner: Arc::new(RwLock::new(Vec::new())),
+            inner: Arc::new(RwLock::new(BTreeSet::new())),
+            submission_number: Arc::new(Mutex::new(0)),
             tx_listeners: Arc::new(Mutex::new(Vec::new())),
             impersonation,
+            transactions_order: Arc::new(RwLock::new(transactions_order)),
```
Comment (dutterbutter) [inline crates/core/src/node/pool.rs:29]:
```suggestion
            transaction_order: Arc::new(RwLock::new(transaction_order)),
```

File: crates/core/src/node/pool.rs [hunk 11]
```diff
@@ -1,59 +1,96 @@
...
     }
 
+    fn lock_submission_number(&self) -> MutexGuard<'_, u64> {
+        self.submission_number
+            .lock()
+            .expect("submission_number lock is poisoned")
+    }
+
+    fn read_transactions_order(&self) -> RwLockReadGuard<'_, TransactionOrder> {
+        self.transactions_order
```
Comment (dutterbutter) [inline crates/core/src/node/pool.rs:40]:
```suggestion
        self.transaction_order
```

**Review by dutterbutter — COMMENTED** [review]
Just the nit with naming convention

**Review by dutterbutter — DISMISSED** [review]
lgtm

**Comment by Romsters** [comment]
Closes [#401](https://github.com/matter-labs/anvil-zksync/issues/401)

**Review by dutterbutter — APPROVED** [review]

**Pull request merged**